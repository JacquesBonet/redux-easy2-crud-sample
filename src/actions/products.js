import PRODUCT from "../constants/products";
import {actions} from "redux-easy2-crud"


export const addProduct = (aProduct) => (actions.create(PRODUCT, aProduct));
export const readProducts = () => (actions.read(PRODUCT));
export const updateProduct = (product) => (actions.update(product));
export const deleteProduct = (product) => (actions.del(product));
