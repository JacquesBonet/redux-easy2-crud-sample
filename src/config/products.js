export const productInfos = {
  FR1: {
    name: 'Fruit Tea',
    prices: [{
      count: -1,
      value: 3.11
    }]
  },
  AP1: {
    name: 'Apple',
    free: ['FR1'],
    prices: [{
      count: 2,
      value: 5.00
    },
    {
      count: -1,
      value: 4.50
    }]
  },
  CF1: {
    name: 'Coffee',
    free: ['FR1'],
    prices: [{
      count: -1,
      value: 11.23
    }]
  }
};


/**
 * Calculate the number of products of type product.productCode
 *
 * @param products       {Array}   Array of products
 * @param code           {String}  The product code to  search products
 * @param position      {int}      The product position on the basket
 * @returns {*}                    return the product number matching productCode code
 */
const getNbProductsWithSameCode = (products, code, position) => {
  return products.reduce((total, product, idx) => {
    if (idx > position) return total;
    if (product.productCode === code) return total + 1;
    return total;
  }, 0)
};

export const getProductCodes = () => (Object.keys(productInfos).sort());

export const getProductName = (productOrCode) => {
  let productCode = productOrCode.productCode ? productOrCode.productCode : productOrCode;
  return productInfos[productCode] ? productInfos[productCode].name : '';
};

/**
 * Calculate the product price
 *
 * @param products      {Array}    Array of products
 * @param product       {Object}   Product to calculate the price
 * @param position      {int}      The product position on the basket
 * @returns {int}                  return the price
 */
export const getProductValue = (products, position = 0) => {
  let product = products[position];
  if (product.hasOwnProperty('price'))  return product.price;

  let nbProducts = getNbProductsWithSameCode( products, product.productCode, position)
  let prices = productInfos[product.productCode].prices;

  for( var price of prices) {
    if (nbProducts <= price.count) return price.value;

    if (price.count === -1) return price.value;
  }
};

/**
 * Return the free products discount associated to a product
 *
 * @param productCode  {Array}    the free product discount. 0, if no product discount
 */
export const hasFreeProductDiscount = (productCode) => {
  const product = productInfos[productCode];

  if (product) return product.free

  return 0;
};


