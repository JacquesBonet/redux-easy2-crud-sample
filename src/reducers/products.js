import PRODUCTS from "../constants/products";
import {crudReducer}  from 'redux-easy2-crud';

const initialState = [];


export default function products(state = initialState, action) {
  return crudReducer(state, PRODUCTS, action);
}
