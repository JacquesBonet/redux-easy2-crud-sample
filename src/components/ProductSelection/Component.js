import React from "react";
import {getProductCodes, getProductName} from "../../config/products";
import AddProductButton from "./AddProductButton";
import "./Component.css";

const ProductSelection = ({productsActions}) => (
  <div className="app__product-selection">
    {
      getProductCodes().map((productCode, idx) => (
        <AddProductButton
          key={idx}
          onClick={() => productsActions.addProduct({productCode})}
          title={getProductName(productCode)}
        />
      ))
    }
  </div>
);

export default ProductSelection;
