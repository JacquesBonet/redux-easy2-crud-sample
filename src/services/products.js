import {getProductCodes, getProductName, getProductValue, hasFreeProductDiscount} from "../config/products";
export {getProductCodes, getProductName, getProductValue, hasFreeProductDiscount};

/**
 * return the total products price in the basket
 *
 * @param productsState         {Array}    the redux state
 * @returns {*}
 */
export const getTotal = (productsState) => {
  return productsState.reduce((total, current, idx) => {
    return total + getProductValue( productsState, idx);
  }, 0);
};


/**
 * return the list of products
 *
 * @param rootState     {Object}    the redux state
 * @returns {Array}
 */
export const getProducts = (rootState) => (rootState.products);


/**
 * id generation
 * @returns {String}
 */
export const id = () => {
  return Math.random().toString(36).substring(7)
}
