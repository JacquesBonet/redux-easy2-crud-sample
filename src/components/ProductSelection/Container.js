import ProductSelection from "./Component";
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import * as productsActions from "../../actions/products";

const mapDispatchToProps = (dispatch) => ({
  productsActions: bindActionCreators(productsActions, dispatch)
});

export default connect(null, mapDispatchToProps)(ProductSelection);
