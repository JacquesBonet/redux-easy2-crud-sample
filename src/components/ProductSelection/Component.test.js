import React from "react";
import {shallow} from "enzyme";
import ProductSelection from "./Component";
import AddProductButton from "./AddProductButton";
import {getProductCodes, getProductName} from "../../config/products";

describe('ProductSelectionComponent', function () {
  describe('render', function () {
    const productsActions = {
      addProduct : jest.fn()
    };
    const productCodes = getProductCodes();
    let component;
    beforeEach(function () {
      component = shallow(<ProductSelection productsActions={productsActions}/>);
    });

    describe('product buttons', function () {
      let productButtons;

      beforeEach(function () {
        productButtons = component.find(AddProductButton);
      });

      it('should have three products', function () {
        expect(productButtons).toHaveLength(productCodes.length);
      });

      productCodes.forEach(function (productCode, idx) {
        let productButton;
        let expectedTitle = getProductName(productCode);
        beforeEach(function () {
          productButton = productButtons.at(idx);
        });

        it(`should have title ${expectedTitle}`, function () {
          expect(productButton.html()).toMatchSnapshot();
        });

        it('should pass the product code to addProduct when clicked', function () {
          productButton.simulate('click');
          expect(productsActions.addProduct).toHaveBeenCalledWith({productCode});
        });
      });
    });
  });
});
