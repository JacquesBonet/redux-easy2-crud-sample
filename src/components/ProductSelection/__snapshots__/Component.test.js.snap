exports[`ProductSelectionComponent render product buttons should have title Apple 1`] = `"<button>Apple</button>"`;

exports[`ProductSelectionComponent render product buttons should have title Coffee 1`] = `"<button>Coffee</button>"`;

exports[`ProductSelectionComponent render product buttons should have title Fruit Tea 1`] = `"<button>Fruit Tea</button>"`;
